File: Chapter3/Python/Example7.py
# Chapter 3 - Binomial Distribution
# Example 7: Comparing Two Production Lines  A~B(50,0.08), B~B(60,0.05)

import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import binom

# Parameters
n_A, p_A = 50, 0.08
n_B, p_B = 60, 0.05

# i. P(X_A = 4 and X_B = 3)  — independent lines
prob_A_4  = binom.pmf(4, n_A, p_A)
prob_B_3  = binom.pmf(3, n_B, p_B)
prob_both = prob_A_4 * prob_B_3
print(f"i. P(X_A=4 and X_B=3) = {prob_both:.6f}")

# ii. P(X_A > 5 or X_B > 5)
prob_A_gt5    = 1 - binom.cdf(5, n_A, p_A)
prob_B_gt5    = 1 - binom.cdf(5, n_B, p_B)
prob_both_gt5 = prob_A_gt5 * prob_B_gt5
prob_union    = prob_A_gt5 + prob_B_gt5 - prob_both_gt5
print(f"\nii. P(X_A>5 or X_B>5) = {prob_union:.6f}")
print(f"    P(X_A>5) = {prob_A_gt5:.6f}")
print(f"    P(X_B>5) = {prob_B_gt5:.6f}")

# iii. Expected total defects and standard deviation
mean_A = n_A * p_A;  mean_B = n_B * p_B
var_A  = n_A * p_A * (1 - p_A);  var_B = n_B * p_B * (1 - p_B)
mean_total = mean_A + mean_B
sd_total   = np.sqrt(var_A + var_B)

print(f"\niii. Expected defects:")
print(f"     Line A: {mean_A:.2f}")
print(f"     Line B: {mean_B:.2f}")
print(f"     Total:  {mean_total:.2f}")
print(f"     Std Dev: {sd_total:.2f}")

# iv. Comparison plots
fig, axes = plt.subplots(1, 2, figsize=(14, 5))

x_A   = np.arange(0, n_A + 1)
pmf_A = binom.pmf(x_A, n_A, p_A)
axes[0].bar(x_A, pmf_A, color='steelblue', alpha=0.7, edgecolor='black')
axes[0].axvline(mean_A, color='red', linestyle='--', linewidth=2,
                label=f'Mean = {mean_A:.2f}')
axes[0].set_xlabel('Number of Defects', fontsize=11)
axes[0].set_ylabel('Probability', fontsize=11)
axes[0].set_title(f'Line A: B({n_A}, {p_A})', fontsize=12)
axes[0].legend()
axes[0].grid(axis='y', alpha=0.3)

x_B   = np.arange(0, n_B + 1)
pmf_B = binom.pmf(x_B, n_B, p_B)
axes[1].bar(x_B, pmf_B, color='coral', alpha=0.7, edgecolor='black')
axes[1].axvline(mean_B, color='red', linestyle='--', linewidth=2,
                label=f'Mean = {mean_B:.2f}')
axes[1].set_xlabel('Number of Defects', fontsize=11)
axes[1].set_ylabel('Probability', fontsize=11)
axes[1].set_title(f'Line B: B({n_B}, {p_B})', fontsize=12)
axes[1].legend()
axes[1].grid(axis='y', alpha=0.3)

plt.tight_layout()
plt.show()

# Overlay comparison
plt.figure(figsize=(12, 6))
plt.plot(x_A, pmf_A, 'o-', color='steelblue', linewidth=2,
         markersize=6, label=f'Line A: B({n_A}, {p_A})')
plt.plot(x_B, pmf_B, 's-', color='coral', linewidth=2,
         markersize=6, label=f'Line B: B({n_B}, {p_B})')
plt.xlabel('Number of Defects', fontsize=12)
plt.ylabel('Probability', fontsize=12)
plt.title('Comparison of Two Production Lines', fontsize=14)
plt.legend(fontsize=11)
plt.grid(alpha=0.3)
plt.tight_layout()
plt.show()

File: Chapter3/R/Example7.R
# Chapter 3 - Binomial Distribution
# Example 7: Comparing Two Production Lines  A~B(50,0.08), B~B(60,0.05)

# Parameters
n_A <- 50;  p_A <- 0.08
n_B <- 60;  p_B <- 0.05

# i. P(X_A = 4 and X_B = 3)  — independent lines
prob_A_4  <- dbinom(4, n_A, p_A)
prob_B_3  <- dbinom(3, n_B, p_B)
prob_both <- prob_A_4 * prob_B_3
cat(sprintf("i. P(X_A=4 and X_B=3) = %.6f\n", prob_both))

# ii. P(X_A > 5 or X_B > 5)
prob_A_gt5    <- 1 - pbinom(5, n_A, p_A)
prob_B_gt5    <- 1 - pbinom(5, n_B, p_B)
prob_both_gt5 <- prob_A_gt5 * prob_B_gt5
prob_union    <- prob_A_gt5 + prob_B_gt5 - prob_both_gt5
cat(sprintf("\nii. P(X_A>5 or X_B>5) = %.6f\n", prob_union))
cat(sprintf("    P(X_A>5) = %.6f\n", prob_A_gt5))
cat(sprintf("    P(X_B>5) = %.6f\n", prob_B_gt5))

# iii. Expected total defects and standard deviation
mean_A    <- n_A * p_A;  mean_B <- n_B * p_B
var_A     <- n_A * p_A * (1 - p_A);  var_B <- n_B * p_B * (1 - p_B)
mean_total <- mean_A + mean_B
sd_total   <- sqrt(var_A + var_B)

cat(sprintf("\niii. Expected defects:\n"))
cat(sprintf("     Line A: %.2f\n", mean_A))
cat(sprintf("     Line B: %.2f\n", mean_B))
cat(sprintf("     Total:  %.2f\n", mean_total))
cat(sprintf("     Std Dev: %.2f\n", sd_total))

# iv. Comparison plots
par(mfrow = c(1, 2))

x_A   <- 0:n_A
pmf_A <- dbinom(x_A, n_A, p_A)
barplot(pmf_A, names.arg = x_A, col = 'steelblue',
        xlab = 'Number of Defects', ylab = 'Probability',
        main = paste0('Line A: B(', n_A, ', ', p_A, ')'),
        border = 'black', las = 1, cex.names = 0.7)
abline(v = mean_A + 0.5, col = 'red', lwd = 2, lty = 2)

x_B   <- 0:n_B
pmf_B <- dbinom(x_B, n_B, p_B)
barplot(pmf_B, names.arg = x_B, col = 'coral',
        xlab = 'Number of Defects', ylab = 'Probability',
        main = paste0('Line B: B(', n_B, ', ', p_B, ')'),
        border = 'black', las = 1, cex.names = 0.7)
abline(v = mean_B + 0.5, col = 'red', lwd = 2, lty = 2)

par(mfrow = c(1, 1))

# Overlay comparison
plot(x_A, pmf_A, type = 'b', pch = 19, col = 'steelblue', lwd = 2,
     xlim = c(0, max(n_A, n_B) * p_A * 3),
     xlab = 'Number of Defects', ylab = 'Probability',
     main = 'Comparison of Two Production Lines')
lines(x_B, pmf_B, type = 'b', pch = 19, col = 'coral', lwd = 2)
legend('topright',
       legend = c(paste('Line A: B(', n_A, ',', p_A, ')'),
                  paste('Line B: B(', n_B, ',', p_B, ')')),
       col = c('steelblue', 'coral'), lwd = 2, pch = 19)
